pinot-query-planner/src/main/java/org/apache/pinot/query/planner/serde/PlanNodeSerializationVisitor.java

```diff
@@ -0,0 +1,386 @@
+import org.apache.pinot.query.planner.plannode.ProjectNode;
+import org.apache.pinot.query.planner.plannode.SetOpNode;
+import org.apache.pinot.query.planner.plannode.SortNode;
+import org.apache.pinot.query.planner.plannode.TableScanNode;
+import org.apache.pinot.query.planner.plannode.ValueNode;
+import org.apache.pinot.query.planner.plannode.WindowNode;
+
+
+public class PlanNodeSerializationVisitor implements PlanNodeVisitor<Void, Plan.StageNode.Builder> {
+  public static final PlanNodeSerializationVisitor INSTANCE = new PlanNodeSerializationVisitor();
+
+  private PlanNodeSerializationVisitor() {
+  }
+
+  public Plan.StageNode process(AbstractPlanNode planNode) {
+    Plan.StageNode.Builder builder = Plan.StageNode.newBuilder().setStageId(planNode.getPlanFragmentId());
+    DataSchema dataSchema = planNode.getDataSchema();
+    for (int i = 0; i < dataSchema.getColumnNames().length; i++) {
+      builder.addColumnNames(dataSchema.getColumnName(i));
+      builder.addColumnDataTypes(RexExpressionToProtoExpression.convertColumnDataType(dataSchema.getColumnDataType(i)));
+    }
+
+    planNode.visit(this, builder);
+    planNode.getInputs().forEach(input -> builder.addInputs(process((AbstractPlanNode) input)));
+
+    return builder.build();
+  }
+
+  @Override
+  public Void visitAggregate(AggregateNode node, Plan.StageNode.Builder builder) {
+    Plan.AggregateNode.Builder aggregateNodeBuilder =
+        Plan.AggregateNode.newBuilder().setNodeHint(getNodeHintBuilder(node.getNodeHint()))
+            .setAggCalls(convertExpressions(node.getAggCalls())).addAllFilterArgIndices(node.getFilterArgIndices())
+            .setGroupSet(convertExpressions(node.getGroupSet())).setAggType(convertAggType(node.getAggType()));
+
+    builder.setAggregateNode(aggregateNodeBuilder);
+    return null;
+  }
+
+  @Override
+  public Void visitFilter(FilterNode node, Plan.StageNode.Builder builder) {
+    Expressions.RexExpression condition = RexExpressionToProtoExpression.process(node.getCondition());
+    Plan.FilterNode.Builder filterNodeBuilder = Plan.FilterNode.newBuilder().setCondition(condition);
+    builder.setFilterNode(filterNodeBuilder);
+
+    return null;
+  }
+
+  @Override
+  public Void visitJoin(JoinNode node, Plan.StageNode.Builder builder) {
+    Plan.JoinKeys.Builder joinKeyBuilder = Plan.JoinKeys.newBuilder().addAllLeftKeys(node.getJoinKeys().getLeftKeys())
+        .addAllRightKeys(node.getJoinKeys().getRightKeys());
+
+    Plan.JoinNode.Builder joinNodeBuilder =
+        Plan.JoinNode.newBuilder().setJoinRelType(convertJoinRelType(node.getJoinRelType())).setJoinKeys(joinKeyBuilder)
+            .setJoinClause(convertExpressions(node.getJoinClauses()))
+            .setJoinHints(getNodeHintBuilder(node.getJoinHints())).addAllLeftColumnNames(node.getLeftColumnNames())
+            .addAllRightColumnNames(node.getRightColumnNames());
+
+    builder.setJoinNode(joinNodeBuilder);
```

Comment:
nit: This is completely subjective, but I find very difficult to read this kind of calls. Don't you find it easier to read when there is one single fluent method call per line? Something like:

```java
Plan.JoinNode.Builder joinNodeBuilder = Plan.JoinNode.newBuilder()
        .setJoinRelType(convertJoinRelType(node.getJoinRelType()))
        .setJoinKeys(joinKeyBuilder)
        .setJoinClause(convertExpressions(node.getJoinClauses()))
        .setJoinHints(getNodeHintBuilder(node.getJoinHints()))
        .addAllLeftColumnNames(node.getLeftColumnNames())
        .addAllRightColumnNames(node.getRightColumnNames());
```

We don't have any code style rule that force use to use one or the other mode, so it is just a nit

Reply:
I did not write the code this way. This is formatted by intellij using the project's coding rules. 
 
 I ran an experiment where I modified the code so that there is one fn. per line as you have pasted above. Then I formatted the file (Code -> Reformat File ...). The editor changed the format back to fitting in as many function calls as possible into a line. It is too much of an hassle to disallow a developer to reformat this file. The best option is to add this rule in the future and then make the change.

Reply:
The intellij convention is not using the project code styles. Or being more precise, the project code style does not define this situation, so intellij is free to do whatever it does by default.

TBH this is not even defined in Google style guide, which basically says to do whatever it considered better. The fact that Intellij just tries to add as many methods as possible in the same line is a cheap decision by their side. I would love that by default intellij would have decided to do not modify these lines given it there is no clear winner in 100% of the cases.

So my suggestion is to disable that option in intellij code style, but again, this is not a strong policy we have in the project.

Reply:
This link says that it does use the coding style setup in the config: https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html

Previously I was told to setup the Pinot code style and reformat the file as the code did not meet some of the standards.